core/backend/src/StashManager.ts

```diff
@@ -0,0 +1,458 @@
+   * @returns The absolute path to the stash root directory.
+   * @throws IModelError if the database is not open, is readonly, has unsaved changes, or if the briefcase path cannot be determined.
+   */
+  private static getStashRootFolder(db: BriefcaseDb, ensureExists: boolean): LocalDirName {
+    if (!db.isOpen || db.isReadonly)
+      throw new StashError("Database is not open or is readonly");
+
+    if (!existsSync(db[_nativeDb].getFilePath())) {
+      throw new StashError("Could not determine briefcase path");
+    }
+
+    const stashDir = path.join(path.dirname(db[_nativeDb].getFilePath()), StashManager.STASH_ROOT_DIR_NAME, `${db.briefcaseId}`);
+    if (ensureExists && !existsSync(stashDir)) {
+      mkdirSync(stashDir, { recursive: true });
+    }
+    return stashDir;
+  }
+
+  /**
+   * Retrieves the stash ID from the provided arguments.
+   *
+   * If the `stash` property of `args` is a string, it returns the string in lowercase.
+   * If the `stash` property is an object, it returns the `id` property of the object in lowercase.
+   *
+   * @param args - The arguments containing the stash information, which can be either a string or an object with an `id` property.
+   * @returns The stash ID as a lowercase string.
+   */
+  private static getStashId(args: StashArgs) {
+    return (typeof args.stash === "string" ? args.stash : args.stash.id).toLowerCase();
+  }
+
+  /**
+   * Retrieves the file path to the stash file associated with the provided arguments.
+   *
+   * @param args - The arguments required to identify the stash, including the database reference.
+   * @returns The absolute path to the stash file.
+   * @throws IModelError with status `IModelStatus.BadArg` if the stash root folder or the stash file does not exist.
+   */
+  private static getStashFilePath(args: StashArgs) {
+    const stashRoot = this.getStashRootFolder(args.db, false);
+    if (!existsSync(stashRoot)) {
+      throw new StashError("Stash root folder does not exist");
+    }
+
+    const stashFile = path.join(stashRoot, `${this.getStashId(args)}.stash`);
+    if (!existsSync(stashFile)) {
+      throw new StashError("Invalid stash");
+    }
+    return stashFile;
+  }
+
+  /**
+   * Queries the stash database for lock IDs matching the specified state and origin.
+   *
+   * @param args - The arguments required to access the stash database.
+   * @param state - The lock state to filter by.
+   * @param origin - The lock origin to filter by.
+   * @returns An array of lock IDs (`Id64Array`) that match the given state and origin.
+   */
+  private static queryLocks(args: StashArgs, state: LockState, origin: LockOrigin): Id64Array {
+    return this.withStash(args, (stashDb) => {
+      const query = `SELECT JSON_GROUP_ARRAY(FORMAT('0x%x', Id)) FROM [locks] WHERE State = ${state} AND origin = ${origin}`;
+      return stashDb.withPreparedSqliteStatement(query, (stmt) => {
+        while (stmt.step() === DbResult.BE_SQLITE_ROW) {
```

Comment:
Does this use `while` instead of `if` because that's the standard for dealing with the results of a SQL fetch? Obviously there will never be more than one iteration of the "loop".

Reply:
if will be fine here as well. It just have single row as result.